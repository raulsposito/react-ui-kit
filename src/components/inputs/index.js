import TextField from './textField'
import Password from './password'
import NumberField from './numberField'
import TextArea from './textArea'

export {
  TextField,
  Password,
  NumberField,
  TextArea
}
