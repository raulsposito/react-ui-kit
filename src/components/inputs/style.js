import styled from 'styled-components'

export const StyledInputWrapper = styled.div`
  width: 100%;

  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  align-items: flex-start;

  position: relative;
`

export const StyledInputWrapperHelper = styled.div`
  width: 100%;

  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: flex-start;
  flex-wrap: wrap;

  position: relative;
`
