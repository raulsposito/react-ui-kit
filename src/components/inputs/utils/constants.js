export const BASIC_INPUT_TYPES = {
  TEXT: 'text',
  EMAIL: 'email',
  TEL: 'tel',
  PASSWORD: 'password',
  NUMBER: 'number',
  NAME: 'name',
  TEXTAREA: 'textarea'
}

export const INPUT_TYPES = {
  ...BASIC_INPUT_TYPES,
  DATE: 'date'
}
