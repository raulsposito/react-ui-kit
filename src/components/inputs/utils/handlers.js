import { INPUT_TYPES } from './constants'

export const handleTextColorState = (error, focused, color) => {
  if (error) return 'error'
  if (focused) return color || 'primary'
  return 'contrast'
}

export const handleTypes = type => {
  switch (type) {
    case INPUT_TYPES.TEXT: return 'text'
    case INPUT_TYPES.NAME: return 'text'
    case INPUT_TYPES.TEL: return 'tel'
    case INPUT_TYPES.EMAIL: return 'email'
    case INPUT_TYPES.NUMBER: return 'number'
    case INPUT_TYPES.PASSWORD: return 'password'
    default: return 'text'
  }
}
