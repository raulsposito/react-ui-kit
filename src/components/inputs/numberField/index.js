import React, { useState } from 'react'
import PropTypes from 'prop-types'

import MyPropTyes from '@global/propTypes'

import { INPUT_TYPES } from '../utils/constants'
import { validateField } from '../utils/validation'
import { StyledInputWrapper } from '../style'

import { handleTextColorState } from '../utils/handlers'

import InputLabel from '../components/inputLabel'
import InputContainer from '../components/inputContainer'
import InputHelperText from '../components/inputHelperText'
import InputAdornment from '../components/inputAdornment'

const NumberField = ({
  name,
  color,
  label,
  helperText,
  onChange,
  placeholder,
  value,
  disabled,
  error,
  required
}) => {
  const [focused, setFocused] = useState(false)
  const [fieldError, setFieldError] = useState(error)

  const handleError = () => {
    const type = INPUT_TYPES.NUMBER
    const inputError = validateField({ type, value, required })

    if (!inputError.isValid) {
      setFieldError(inputError.message)
    } else {
      setFieldError(error)
    }
  }

  const handleFocus = () => {
    setFocused(true)
  }

  const handleBlur = () => {
    setFocused(false)
    handleError()
  }

  const handleChange = inputValue => onChange(name, Number(inputValue))
  const incrementValue = () => handleChange(Number(value) + 1)
  const decreaseValue = () => handleChange(Number(value) - 1)

  return (
    <StyledInputWrapper>
      <InputLabel
        text={label}
        color={handleTextColorState(fieldError, focused, color)}
        name={name}
        disabled={disabled}
        error={fieldError}
        required={required}
      />
      <InputContainer
        type='number'
        value={value}
        placeholder={placeholder}
        onChange={e => handleChange(e.target.value)}
        onFocus={handleFocus}
        onBlur={handleBlur}
        name={name}
        disabled={disabled}
        error={fieldError}
        focused={focused}
        required={required}
        color={color}
        startAdornment={(
          <InputAdornment
            onClick={!disabled && decreaseValue}
            position='start'
            disabled={disabled}
            icon='expand_more'
            color={fieldError ? 'error' : 'system'}
          />
            )}
        endAdornment={(
          <InputAdornment
            onClick={!disabled && incrementValue}
            position='end'
            disabled={disabled}
            icon='expand_less'
            color={fieldError ? 'error' : 'system'}
          />
          )}
      />
      <InputHelperText
        text={fieldError || helperText}
        color={handleTextColorState(fieldError, focused, color)}
        disabled={disabled}
      />
    </StyledInputWrapper>
  )
}

NumberField.propTypes = {
  name: PropTypes.string,
  color: MyPropTyes.color,
  label: PropTypes.string,
  helperText: PropTypes.string,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  value: PropTypes.number,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  required: PropTypes.bool
}

export default NumberField
