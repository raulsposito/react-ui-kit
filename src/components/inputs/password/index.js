import React, { useState } from 'react'
import PropTypes from 'prop-types'

import MyPropTyes from '@global/propTypes'

import { INPUT_TYPES } from '../utils/constants'
import { validateField } from '../utils/validation'
import { StyledInputWrapper } from '../style'

import { handleTextColorState } from '../utils/handlers'

import InputLabel from '../components/inputLabel'
import InputContainer from '../components/inputContainer'
import InputHelperText from '../components/inputHelperText'
import InputAdornment from '../components/inputAdornment'

const Password = ({ name, color, label, helperText, onChange, placeholder, value, disabled, error, required }) => {
  const [focused, setFocused] = useState(false)
  const [passwordVisibility, setPasswordVisibility] = useState(false)
  const [fieldError, setFieldError] = useState(error)

  const handleError = () => {
    const type = INPUT_TYPES.PASSWORD
    const inputError = validateField({ type, value, required })

    if (!inputError.isValid) {
      setFieldError(inputError.message)
    } else {
      setFieldError(error)
    }
  }

  const handleFocus = () => {
    setFocused(true)
  }

  const handleBlur = () => {
    setFocused(false)
    handleError()
  }

  const handleChange = e => onChange(name, e.target.value)

  const handlePasswordVisibility = () => {
    setPasswordVisibility(!passwordVisibility)
  }

  return (
    <StyledInputWrapper>
      <InputLabel
        text={label}
        color={handleTextColorState(fieldError, focused, color)}
        name={name}
        disabled={disabled}
        error={fieldError}
        focused={focused}
        required={required}
      />
      <InputContainer
        type={passwordVisibility ? 'text' : 'password'}
        value={value}
        placeholder={placeholder}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        name={name}
        disabled={disabled}
        error={fieldError}
        focused={focused}
        required={required}
        color={color}
        endAdornment={(
          <InputAdornment
            onClick={!disabled ? handlePasswordVisibility : null}
            position='end'
            disabled={disabled}
            icon={!passwordVisibility ? 'visibility' : 'visibility_off'}
            color={fieldError ? 'error' : 'system'}
          />
        )}
      />
      <InputHelperText
        text={fieldError || helperText}
        color={handleTextColorState(fieldError, focused, color)}
        disabled={disabled}
        error={fieldError}
        focused={focused}
      />
    </StyledInputWrapper>
  )
}

Password.propTypes = {
  name: PropTypes.string,
  color: MyPropTyes.color,
  label: PropTypes.string,
  helperText: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.any,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  required: PropTypes.bool
}

export default Password
