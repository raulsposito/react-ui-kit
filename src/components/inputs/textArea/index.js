import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'

import MyPropTyes from '@global/propTypes'

import { StyledInputWrapper, StyledInputWrapperHelper } from '../style'

import { handleTextColorState } from '../utils/handlers'

import InputLabel from '../components/inputLabel'
import InputHelperText from '../components/inputHelperText'
import InputCharacterCounter from '../components/inputCharacterCounter'
import InputAreaContainer from '../components/inputAreaContainer'
import { validateTextArea } from '../utils/validation'

const TextArea = props => {
  const {
    name,
    rows,
    color,
    label,
    value,
    error,
    resize,
    onChange,
    disabled,
    required,
    maxLength,
    helperText,
    placeholder
  } = props

  const [focused, setFocused] = useState(false)
  const [fieldError, setFieldError] = useState(error)
  const [characterCount, setCharacterCount] = useState(value.length)

  useEffect(() => {
    setCharacterCount(value.length)
  }, [value])

  const handleError = () => {
    const inputError = validateTextArea({ value, maxLength, required })

    if (!inputError.isValid) {
      setFieldError(inputError.message)
    } else {
      setFieldError(error)
    }
  }

  const handleFocus = () => {
    setFocused(true)
  }

  const handleBlur = () => {
    setFocused(false)
    handleError()
  }

  const handleChange = e => onChange(name, e.target.value)

  return (
    <StyledInputWrapper>
      <InputLabel
        text={label}
        color={handleTextColorState(fieldError, focused, color)}
        name={name}
        disabled={disabled}
        error={fieldError}
        required={required}
      />

      <InputAreaContainer
        value={value}
        name={name}
        placeholder={placeholder}
        required={required}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        disabled={disabled}
        error={fieldError}
        maxLength={maxLength}
        rows={rows}
        resize={resize}
      />
      <StyledInputWrapperHelper>
        <InputCharacterCounter
          currentCharacterCount={characterCount}
          maxLength={maxLength}
          disabled={disabled}
        />
        <InputHelperText
          text={fieldError || helperText}
          color={handleTextColorState(fieldError, focused, color)}
          disabled={disabled}
        />
      </StyledInputWrapperHelper>
    </StyledInputWrapper>
  )
}

TextArea.propTypes = {
  name: PropTypes.string,
  color: MyPropTyes.color,
  label: PropTypes.string,
  helperText: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.any,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  required: PropTypes.bool,
  maxLength: PropTypes.number,
  rows: PropTypes.number,
  resize: PropTypes.bool
}

export default TextArea
