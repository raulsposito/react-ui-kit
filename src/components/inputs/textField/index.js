import React, { useState } from 'react'
import PropTypes from 'prop-types'

import MyPropTyes from '@global/propTypes'

import { BASIC_INPUT_TYPES } from '@components/inputs/utils/constants'
import { validateField } from '../utils/validation'
import { StyledInputWrapper } from '../style'

import { handleTextColorState } from '../utils/handlers'

import InputLabel from '../components/inputLabel'
import InputContainer from '../components/inputContainer'
import InputHelperText from '../components/inputHelperText'
import InputAdornment from '../components/inputAdornment'

const TextField = ({ name, color, label, helperText, onChange, placeholder, type, value, endAdornment, startAdornment,
  disabled, error, required }) => {
  const [focused, setFocused] = useState(false)
  const [fieldError, setFieldError] = useState(error)
  const errorAdornment = <InputAdornment position='end' icon='error' color='error' />

  const handleError = () => {
    const inputError = validateField({ type, value, required })

    if (!inputError.isValid) {
      setFieldError(inputError.message)
    } else {
      setFieldError(error)
    }
  }

  const handleFocus = () => {
    setFocused(true)
  }

  const handleBlur = () => {
    setFocused(false)
    handleError()
  }

  const handleChange = e => onChange(name, e.target.value)

  return (
    <StyledInputWrapper>
      <InputLabel
        text={label}
        color={handleTextColorState(fieldError, focused, color)}
        name={name}
        disabled={disabled}
        error={fieldError}
        focused={focused}
        required={required}
      />
      <InputContainer
        type={type}
        value={value}
        placeholder={placeholder}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        name={name}
        startAdornment={startAdornment}
        endAdornment={fieldError ? errorAdornment : endAdornment}
        disabled={disabled}
        error={fieldError}
        focused={focused}
        required={required}
        color={color}
      />
      <InputHelperText
        text={fieldError || helperText}
        color={handleTextColorState(fieldError, focused, color)}
        disabled={disabled}
        error={fieldError}
        focused={focused}
      />
    </StyledInputWrapper>
  )
}

TextField.propTypes = {
  name: PropTypes.string,
  color: MyPropTyes.color,
  label: PropTypes.string,
  helperText: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  type: PropTypes.oneOf(Object.values(BASIC_INPUT_TYPES)),
  value: PropTypes.any,
  endAdornment: PropTypes.node,
  startAdornment: PropTypes.node,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  required: PropTypes.bool
}

export default TextField
