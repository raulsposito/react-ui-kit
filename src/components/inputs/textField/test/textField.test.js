import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { colors } from '@global/theme'
import { INPUT_TYPES } from '@components/inputs/utils/constants'

import InputAdornment from '@components/inputs/components/inputAdornment'

import { TextField } from '@components/inputs'

const startAdornment = <InputAdornment icon='visibility' color='system' />
const endAdornment = <InputAdornment icon='help' color='system' />

let value = 'textInsertValue'
const handleChange = e => { value = e }

const mockData = {
  name: 'txtMock',
  label: 'TextField Label',
  requiredLabel: 'TextField Label *',
  placeholder: 'textfField placeholder',
  value: 'goblinShaman',
  error: 'Error Text',
  helperText: 'Helper Text',
  color: 'secondary',
  onChange: jest.fn(),
  startAdornment,
  endAdornment
}

const mockValidationsData = {
  email: {
    correct: 'jumpierrez@shaman.uy',
    wrong: 'jumpierrez'
  },
  name: {
    correct: 'Joaquin',
    wrong: 'Joaquin1'
  }
}

describe('<TextField />', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<TextField
      name={mockData.name}
      label={mockData.label}
      placeholder={mockData.placeholder}
      helperText={mockData.helperText}
    />)
    expect(asFragment()).toMatchSnapshot()
  })

  describe('Disabled', () => {
    it('is disabled', () => {
      render(
        <TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          disabled
        />
      )

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      expect(input).toBeDisabled()
    })
  })

  describe('Required', () => {
    it('is required', () => {
      render(
        <TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          required
        />
      )

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      const label = screen.queryByText(mockData.requiredLabel)

      expect(input).toBeRequired()
      expect(label).toBeInTheDocument()
    })
  })

  describe('Error', () => {
    it('has error (icon)', () => {
      render(<TextField
        name={mockData.name}
        label={mockData.label}
        placeholder={mockData.placeholder}
        helperText={mockData.helperText}
        error={mockData.error}
      />)

      const errorAdornment = screen.queryByText('error')
      expect(errorAdornment).toBeInTheDocument()
    })
  })

  describe('Placeholder', () => {
    it('input has placeholder', () => {
      render(
        <TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
        />
      )

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      expect(input).toBeInTheDocument()
    })
  })

  describe('Name', () => {
    it('input has name', () => {
      render(
        <TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
        />
      )

      const input = screen.queryByPlaceholderText(mockData.placeholder, { name: mockData.name })
      const label = screen.queryByText(mockData.label).closest('label')

      expect(input).toBeInTheDocument()
      expect(label).toHaveAttribute('for', mockData.name)
    })
  })

  describe('Input event', () => {
    it('calls function on change', () => {
      render(<TextField
        name={mockData.name}
        label={mockData.label}
        placeholder={mockData.placeholder}
        helperText={mockData.helperText}
        onChange={mockData.onChange}
      />)

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      fireEvent.change(input, { target: { value: 'new value' } })
      expect(mockData.onChange).toHaveBeenCalledTimes(1)
    })
  })

  describe('Input Type', () => {
    it('no type', () => {
      render(
        <TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
        />
      )

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      expect(input).toHaveAttribute('type', 'text')
    })

    it('is name type', () => {
      render(
        <TextField
          type={INPUT_TYPES.NAME}
          name={mockData.name}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
        />
      )

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      expect(input).toHaveAttribute('type', 'text')
    })

    it('is email type', () => {
      render(
        <TextField
          type={INPUT_TYPES.EMAIL}
          name={mockData.name}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
        />
      )

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      expect(input).toHaveAttribute('type', 'email')
    })
  })

  describe('Handle State Color', () => {
    describe('Handle one state at time', () => {
      it('has not state changes', () => {
        render(
          <TextField
            name={mockData.name}
            label={mockData.label}
            placeholder={mockData.placeholder}
            helperText={mockData.helperText}
          />
        )

        const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')
        const label = screen.queryByText(mockData.label)
        const helper = screen.queryByText(mockData.helperText)

        expect(inputContainer).toHaveStyle({ borderColor: `${colors.system.regular}` })
        expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.regular}` })
        expect(label).toHaveStyle({ color: `${colors.contrast.regular}` })
        expect(helper).toHaveStyle({ color: `${colors.contrast.regular}` })
      })

      it('is disabled', () => {
        render(
          <TextField
            name={mockData.name}
            label={mockData.label}
            placeholder={mockData.placeholder}
            helperText={mockData.helperText}
            disabled
          />
        )

        const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')
        expect(inputContainer).toHaveStyle({ borderColor: `${colors.system.disabled}` })
        expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.disabled}` })
      })

      it('has error', () => {
        render(<TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          error={mockData.error}
        />)

        const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')
        const label = screen.queryByText(mockData.label)
        const helper = screen.queryByText(mockData.error)

        expect(inputContainer).toHaveStyle({ borderColor: `${colors.error.regular}` })
        expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.regular}` })
        expect(label).toHaveStyle({ color: `${colors.error.regular}` })
        expect(helper).toBeInTheDocument()
        expect(helper).toHaveStyle({ color: `${colors.error.regular}` })
      })

      it('has custom color', () => {
        render(<TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          color={mockData.color}
        />)

        const input = screen.queryByPlaceholderText(mockData.placeholder)
        const label = screen.queryByText(mockData.label)
        const helper = screen.queryByText(mockData.helperText)

        fireEvent.focus(input)

        expect(helper).toHaveStyle({ color: `${colors.secondary.regular}` })
        expect(label).toHaveStyle({ color: `${colors.secondary.regular}` })
      })
    })

    describe('Handle multiple states at time', () => {
      it('is disabled and has custom color', () => {
        render(<TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          disabled
          color={mockData.color}
        />)

        const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')

        expect(inputContainer).toHaveStyle({ borderColor: `${colors.system.disabled}` })
        expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.disabled}` })
      })

      it('has error and custom color', () => {
        render(<TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          error={mockData.error}
          color={mockData.color}
        />)

        const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')
        const label = screen.queryByText(mockData.label)
        const helper = screen.queryByText(mockData.error)

        expect(inputContainer).toHaveStyle({ borderColor: `${colors.error.regular}` })
        expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.regular}` })
        expect(label).toHaveStyle({ color: `${colors.error.regular}` })
        expect(helper).toHaveStyle({ color: `${colors.error.regular}` })
      })
    })
  })

  describe('Adornment', () => {
    it('has no adornment', () => {
      render(
        <TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
        />
      )

      const startAdornmentDiv = screen.queryByText('visibility')?.closest('div')
      const endAdornmentDiv = screen.queryByText('help')?.closest('div')

      expect(startAdornmentDiv).toBeUndefined()
      expect(endAdornmentDiv).toBeUndefined()
    })

    it('has start adornment only', () => {
      render(<TextField
        name={mockData.name}
        label={mockData.label}
        placeholder={mockData.placeholder}
        helperText={mockData.helperText}
        startAdornment={startAdornment}
      />)

      const startAdornmentDiv = screen.queryByText('visibility')?.closest('div')
      const endAdornmentDiv = screen.queryByText('help')?.closest('div')

      expect(startAdornmentDiv).toBeInTheDocument()
      expect(endAdornmentDiv).toBeUndefined()
    })

    it('has end adornment only', () => {
      render(
        <TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          endAdornment={mockData.endAdornment}
        />
      )

      const startAdornmentDiv = screen.queryByText('visibility')?.closest('div')
      const endAdornmentDiv = screen.queryByText('help')?.closest('div')

      expect(startAdornmentDiv).toBeUndefined()
      expect(endAdornmentDiv).toBeInTheDocument()
    })

    it('has start and end adornments at the same time', () => {
      render(<TextField
        name={mockData.name}
        label={mockData.label}
        placeholder={mockData.placeholder}
        startAdornment={mockData.startAdornment}
        endAdornment={mockData.endAdornment}
      />)

      const startAdornmentDiv = screen.queryByText('visibility')?.closest('div')
      const endAdornmentDiv = screen.queryByText('help')?.closest('div')

      expect(startAdornmentDiv).toBeInTheDocument()
      expect(endAdornmentDiv).toBeInTheDocument()
    })
  })

  describe('Validations', () => {
    describe('Name', () => {
      it('validation wrong', () => {
        render(<TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          type={INPUT_TYPES.NAME}
          value={mockValidationsData.name.wrong}
        />)

        const input = screen.queryByPlaceholderText(mockData.placeholder)
        const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')
        const label = screen.queryByText(mockData.label)
        const helper = screen.queryByText(mockData.helperText)

        fireEvent.blur(input)
        expect(inputContainer).toHaveStyle({ borderColor: `${colors.error.regular}` })
        expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.regular}` })
        expect(label).toHaveStyle({ color: `${colors.error.regular}` })
        expect(helper).not.toBe(mockData.helperText)
        expect(helper).toHaveStyle({ color: `${colors.error.regular}` })
        expect(helper).toBeInTheDocument()
      })
    })

    describe('Email', () => {
      it('validation wrong', () => {
        render(<TextField
          name={mockData.name}
          label={mockData.label}
          placeholder={mockData.placeholder}
          helperText={mockData.helperText}
          type={INPUT_TYPES.EMAIL}
          value={mockValidationsData.email.wrong}
        />)

        const input = screen.queryByPlaceholderText(mockData.placeholder)
        const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')
        const label = screen.queryByText(mockData.label)
        const helper = screen.queryByText(mockData.helperText)

        fireEvent.blur(input)
        expect(inputContainer).toHaveStyle({ borderColor: `${colors.error.regular}` })
        expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.regular}` })
        expect(label).toHaveStyle({ color: `${colors.error.regular}` })
        expect(helper).not.toBe(mockData.helperText)
        expect(helper).toHaveStyle({ color: `${colors.error.regular}` })
        expect(helper).toBeInTheDocument()
      })
    })

    it('priorize textField validation error from error property', () => {
      render(<TextField
        name={mockData.name}
        label={mockData.label}
        placeholder={mockData.placeholder}
        helperText={mockData.helperText}
        type={INPUT_TYPES.EMAIL}
        value={mockValidationsData.email.wrong}
        error={mockData.error}
      />)

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      const inputContainer = screen.queryByPlaceholderText(mockData.placeholder).closest('div')
      const label = screen.queryByText(mockData.label)
      const helper = screen.queryByText(mockData.error)

      fireEvent.blur(input)
      expect(inputContainer).toHaveStyle({ borderColor: `${colors.error.regular}` })
      expect(inputContainer).toHaveStyle({ backgroundColor: `${colors.background.regular}` })
      expect(label).toHaveStyle({ color: `${colors.error.regular}` })
      expect(helper).not.toBe(mockData.helperText)
      expect(helper).not.toBe(mockData.error)
      expect(helper).toHaveStyle({ color: `${colors.error.regular}` })
      expect(helper).toBeInTheDocument()
    })
  })

  describe('Controlled Component Behavior', () => {
    it('controlled value set', () => {
      render(<TextField
        name={mockData.name}
        label={mockData.label}
        placeholder={mockData.placeholder}
        helperText={mockData.helperText}
        type={INPUT_TYPES.EMAIL}
        error={mockData.error}
        value={value}
        onChange={handleChange}
      />)

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      expect(input.value).toBe(value)
    })

    it('controlled value change', () => {
      render(<TextField
        name={mockData.name}
        label={mockData.label}
        placeholder={mockData.placeholder}
        helperText={mockData.helperText}
        type={INPUT_TYPES.EMAIL}
        error={mockData.error}
        value={value}
        onChange={handleChange}
      />)

      const input = screen.queryByPlaceholderText(mockData.placeholder)
      fireEvent.change(input, { target: { value: 23 } })
      waitFor(() => expect(input.value).toBe(value))
    })
  })
})
