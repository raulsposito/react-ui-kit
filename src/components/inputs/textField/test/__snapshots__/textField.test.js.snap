// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-pGacB dHCwQs"
  >
    <label
      class="sc-kEjbQP egbqa-d"
      for="txtMock"
    >
      <p
        class="sc-bdfBQB bdAWqY"
        color="contrast"
      >
        TextField Label 
      </p>
    </label>
    <div
      class="sc-crrszt bmxvWD"
    >
      <input
        class="sc-iqHYmW fVDaxC"
        name="txtMock"
        placeholder="textfField placeholder"
        type="text"
        value=""
      />
    </div>
    <span
      class="sc-dQoVA xwLTc"
    >
      <p
        class="sc-bdfBQB bdAWqY"
        color="contrast"
      >
        Helper Text
      </p>
    </span>
  </div>
</DocumentFragment>
`;
